Sources/StreamChat/StateLayer/ChannelListState+Observer.swift

```diff
@@ -0,0 +1,146 @@
+                        try await self?.linkChannelIfNeeded(event.channel)
+                    }
+                ),
+                EventObserver(
+                    notificationCenter: nc,
+                    transform: { $0 as? NotificationMessageNewEvent },
+                    callback: { [weak self] event in
+                        try await self?.linkChannelIfNeeded(event.channel)
+                    }
+                ),
+                EventObserver(
+                    notificationCenter: nc,
+                    transform: { $0 as? ChannelUpdatedEvent },
+                    callback: { [weak self] event in
+                        try await self?.unlinkChannelIfNeeded(event.channel)
+                    }
+                ),
+                EventObserver(
+                    notificationCenter: nc,
+                    transform: { $0 as? ChannelVisibleEvent },
+                    callback: { [weak self] event in
+                        guard let self else { return }
+                        let channel = try await self.database.read { context in
+                            guard let dto = ChannelDTO.load(cid: event.cid, context: context) else {
+                                throw ClientError.ChannelDoesNotExist(cid: event.cid)
+                            }
+                            return try dto.asModel()
+                        }
+                        try await self.linkChannelIfNeeded(channel)
+                    }
+                )
+            ]
+            
+            do {
+                return try channelListObserver.startObserving(didChange: handlers.channelsDidChange)
+            } catch {
+                log.error("Failed to start the channel list observer for query: \(query)")
+                return StreamCollection([])
+            }
+        }
+        
+        // MARK: Linking and Unlinking Channels from the Query
+        
+        private func isBelongingToChannelListQuery(channel: ChatChannel) -> Bool {
+            if let filter = dynamicFilter {
+                return filter(channel)
+            }
+            // When auto-filtering is enabled the channel will appear or not automatically if the
+            // query matches the DB Predicate. So here we default to saying it always belong to the current query.
+            if clientConfig.isChannelAutomaticFilteringEnabled {
+                return true
+            }
+            return false
+        }
+        
+        private func isChannelInList(_ cid: ChannelId) async throws -> Bool {
+            try await database.read { [query] context in
+                guard let (channelDTO, queryDTO) = context.getChannelWithQuery(cid: cid, query: query) else { return false }
+                return queryDTO.channels.contains(channelDTO)
+            }
+        }
+        
+        private func linkChannelIfNeeded(_ channel: ChatChannel) async throws {
+            let listContainsChannel = try await isChannelInList(channel.cid)
```

Comment:
this code is also duplicated, right? Let's discuss if/how should we solve this.

Reply:
https://github.com/GetStream/stream-chat-swift/pull/3190

Reply:
Done